src/pow.rs

```diff
@@ -0,0 +1,1314 @@
+        // that much but we define the inverse of 0 as MAX.
+        if *self == Uint256::ZERO {
+            return Uint256::MAX;
+        }
+
+        // We define the inverse of max as 1.
+        if *self == Uint256::MAX {
+            return Uint256::ONE;
+        }
+
+        // And the inverse of 1 as max.
+        if *self == Uint256::ONE {
+            return Uint256::MAX;
+        }
+
+        let mut denom = *self;
+        denom.increment();
+
+        let mut ret = !*self / denom;
+        ret.increment();
+
+        ret
+    }
+
+    /// Returns the low 32 bits.
+    fn low_u32(&self) -> u32 {
+        let &Uint256(ref arr) = self;
+        arr[0] as u32
+    }
+
+    /// Returns the low 64 bits.
+    fn low_u64(&self) -> u64 {
+        let &Uint256(ref arr) = self;
+        arr[0] as u64
+    }
+
+    /// Returns the least number of bits needed to represent the number.
+    fn bits(&self) -> usize {
+        let arr = self.0;
+        for i in 1..4 {
+            if arr[4 - i] > 0 {
+                return (0x40 * (4 - i + 1)) - arr[4 - i].leading_zeros() as usize;
+            }
+        }
+        0x40 - arr[0].leading_zeros() as usize
+    }
+
+    /// Wrapping multiplication by `u32`.
+    fn mul_u32(self, rhs: u32) -> Uint256 {
+        let Uint256(ref arr) = self;
+        let mut carry = [0u64; 4];
+        let mut ret = [0u64; 4];
+        for i in 0..4 {
+            let not_last_word = i < 4 - 1;
+            let upper = rhs as u64 * (arr[i] >> 32);
+            let lower = rhs as u64 * (arr[i] & 0xFFFFFFFF);
+            if not_last_word {
+                carry[i + 1] += upper >> 32;
+            }
+            let (sum, overflow) = lower.overflowing_add(upper << 32);
+            ret[i] = sum;
+            if overflow && not_last_word {
+                carry[i + 1] += 1;
+            }
+        }
+        Uint256(ret).wrapping_add(Uint256(carry))
+    }
+
+    /// Creates a [`Uint256`] from an array using big-endian encoding.
+    fn from_be(array: [u64; 4]) -> Uint256 {
+        let mut res = [0_u64; 4];
+        for (i, b) in array.iter().rev().enumerate() {
+            res[i] = *b;
+        }
+        Uint256(res)
+    }
+
+    /// Creates a [`Uint256`] from an array using little-endian encoding.
+    fn from_le(array: [u64; 4]) -> Uint256 { Uint256(array) }
+
+    /// Creates a [`Uint256`] from a byte array using big-endian encoding.
+    fn from_be_bytes(bytes: [u8; 32]) -> Uint256 { Self::_from_be_slice(&bytes) }
+
+    /// Creates a [`Uint256`] from a byte array using little-endian encoding.
+    fn from_le_bytes(bytes: [u8; 32]) -> Uint256 { Self::_from_le_slice(&bytes) }
+
+    /// Creates a [`Uint256`] from a byte slice using big-endian encoding.
+    fn from_be_slice(bytes: &[u8]) -> Result<Uint256, ParseLengthError> {
+        if bytes.len() != 32 {
+            Err(ParseLengthError { actual: bytes.len(), expected: 32 })
+        } else {
+            Ok(Self::_from_be_slice(bytes))
+        }
+    }
```

Comment:
I'd prefer to remove `from_le_slice` and `from_be_slice` from everywhere. People can use `try_into()` if they have a slice argument and handle the error as neeeded.

Reply:
Removed. While refactoring and unit testing I removed all usage of the `endian` module. I rekon we may be able to remove the module all together. Will investigate as a separate PR.